internal/rootcoord/root_coord.go

```diff
@@ -1048,6 +1051,201 @@ func (c *Core) DropCollection(ctx context.Context, in *milvuspb.DropCollectionRe
+				Reason:    fmt.Sprintf("truncate collection: %s.%s fail while load temp collection, clear is needed before you retry", in.DbName, in.CollectionName),
+			}, nil
+		}
+	}
+
+	if err := c.scheduler.AddTask(t); err != nil {
+		log.Info("failed to enqueue request to exchange collection with its temp", zap.Error(err), zap.Uint64("ts", t.GetTs()))
+
+		metrics.RootCoordDDLReqCounter.WithLabelValues("TruncateCollection", metrics.FailLabel).Inc()
+		return merr.Status(err), nil
+	}
+	if err := t.WaitToFinish(); err != nil {
+		log.Info("failed to exchange collection with its temp", zap.Error(err), zap.Uint64("ts", t.GetTs()))
+		metrics.RootCoordDDLReqCounter.WithLabelValues("TruncateCollection", metrics.FailLabel).Inc()
+		return merr.Status(err), nil
+	}
+
+	metrics.RootCoordDDLReqCounter.WithLabelValues("TruncateCollection", metrics.SuccessLabel).Inc()
+	metrics.RootCoordDDLReqLatency.WithLabelValues("TruncateCollection").Observe(float64(tr.ElapseSpan().Milliseconds()))
+	metrics.RootCoordDDLReqLatencyInQueue.WithLabelValues("TruncateCollection").Observe(float64(t.queueDur.Milliseconds()))
+
+	log.Info("done to truncate collection", zap.Uint64("ts", t.GetTs()))
+	return merr.Success(), nil
+}
+
+// drop temp collection
+// todo: while start a rootcoord
+func (c *Core) dropCollectionForTempSync(ctx context.Context, in *rootcoordpb.TruncateCollectionRequest) (*commonpb.Status, error) {
+	t := &markTempCollectionAsDeleteTask{
+		tempCollectionTask: tempCollectionTask{
+			baseTask:    newBaseTask(ctx, c),
+			TruncateReq: in,
+		},
+	}
+	log := log.Ctx(ctx).With(zap.String("role", typeutil.RootCoordRole),
+		zap.String("dbName", in.GetDbName()), zap.String("collectionName", in.GetCollectionName()),
+		zap.Bool("needLoad", in.NeedLoad), zap.Bool("clearTempCollection", in.Clear), zap.Uint64("ts", t.GetTs()))
+	if err := c.scheduler.AddTask(t); err != nil {
+		log.Info("failed to enqueue request to drop temp collection", zap.Error(err))
+		return merr.Status(err), nil
+	}
+	if err := t.WaitToFinish(); err != nil {
+		log.Info("failed to make temp collection as deleted", zap.Error(err))
+		return merr.Status(err), nil
+	}
+
+	res, err := c.queryCoord.ReleaseCollection(ctx, &querypb.ReleaseCollectionRequest{
+		Base:         commonpbutil.NewMsgBase(commonpbutil.WithMsgType(commonpb.MsgType_ReleaseCollection)),
+		CollectionID: t.CollectionIDs.CollectionID,
+	})
+	if err == nil {
+		err = merr.Error(res)
+	}
+	if err != nil {
+		log.Info("failed to release temp collection", zap.Error(err))
+		return merr.Status(err), nil
+	}
+	res, err = c.dataCoord.DropIndexesForTemp(ctx, &indexpb.CollectionWithTempRequest{
+		TempCollectionID: t.CollectionIDs.TempCollectionID,
+		CollectionID:     t.CollectionIDs.CollectionID,
+	})
+	if err == nil {
+		err = merr.Error(res)
+	}
+	if err != nil {
+		log.Info("failed to release temp collection", zap.Error(err))
+		return merr.Status(err), nil
+	}
+	_, err = c.garbageCollector.GcCollectionData(ctx, t.collInfo)
```

Comment:
Wati for this ts to be synced.

Reply:
In fact, we can make this a redo task, which is very useful to recycle the resources when any error is encounted, especially for releasing collection.

Reply:
Just similar to dropping a collection.